### ISSUE TITLE
Question regarding QM/MM with custom wavefunction file

### ISSUE THREAD

**Issue by Lisa-Phan** [RagnarB83/ash#426]
Good evening, 

My apologies for posting this under issues, since this is more of a question. I'm wondering if ASH's QM/MM has a workflow that supports custom QM/MM with broken symmetry wavefunction, similar to the TCL-ChemShell ORCA developed by the lab previously.

https://github.com/RagnarB83/chemshell-QMMM-protein-setup

Thank you, 
Lisa Phan

**Comment by RagnarB83** [comment]
Hi Lisa

Github issues are fine for questions as well.

Yes absolutely, this is quite straightforward. Roughly anything that is possible in a QM-code interface in ASH should be possible at the QM/MM level.
I realize there was a broken-symmetry DFT example and explanation missing but you can now find it here:
https://ash.readthedocs.io/en/latest/ORCA-interface.html#broken-symmetry-dft-example

This information would apply directly to a QM/MM setup involving ORCA.

**Comment by Lisa-Phan** [comment]
Thank you Ragnar for the syntax guide. That was very helpful!

I have some follow up questions on modifying the system before starting. After going through ASH's tutorial on QMMM, I noticed that there are some helper codes to index, add, or remove atoms of a system for CHARMM. For adding molecules, there are checks that the newly added molecule have their topology already defined in the CHARMM topology file. In the case where the new atoms added belong to a non-standard molecule, like a substrate, the topology file should be parameterized first with an outside software? 

In the case where the molecule strictly belongs in the QM region, would the force constants needed for MD parameters still be required?

Lastly, I would really appreciate advice on how to proceed with a slightly convoluted missing atom case. I have an MD equilibrated system in AMBER that I would like to bring to ASH for QM/MM. The system was built with AMBER because MCPB.py protocol is needed for the metal binding active site. The system has a gas substrate that is missing, so I would have to add it before running ASH. Unfortunately, there are no straightforward protocols to directly add this substrate in AMBER tleap. I'd tried to port AMBER to CHARMM files to test ASH's built-in add atom functions, but there are no good ways to build all files needed for input. 

Thank you, 
Lisa Phan

**Comment by RagnarB83** [comment]
So if you are using different forcefield files then things behave a little differently:

If you use CHARMM FF files (meaning CHARMM PSF file, CHARMM topology and CHARMM parameter file):
- then if you want to change the system (remove or add atoms) then you have to make sure to update the fairly complicated PSF file correctly. For the CHARMM topology (.rtf) and parameter (.prm) files you only have to add new residues and parameters if you have added something new (they typically contain the whole biomolecular forcefield). ASH does feature functions *remove_atoms_from_system_CHARMM* and *add_atoms_to_system_CHARMM* that can help you automate this although most of the tricky part is done by the PSFgen program (for modifying the PSF-file) that you need to download separately. You also have to figure the FF parameters on your own. It may also be better to do this using whatever program you used to create the CHARMM setup in the first place.

If you want to change a system described by Amber FF files (meaning the PRMTOP file), i.e. add or remove atoms then you have to modify the whole PRMTOP file. ASH currently can not help you with this. Modifications are probably best done using Ambertools.

I personally recommend avoiding these old-school formats if you can and set up a new system using the OpenMM XML format as everything becomes easier with them, as the topology is defined via the PDB-file and all forcefield parameters (CHARMM or Amber) are defined within the well-formatted XML files (and you can just create new ones or add to one of them). See e.g rubredoxin tutorial: https://ash.readthedocs.io/en/latest/Metalloprotein-I.html
There is also an option to parameterize new small molecules, see https://ash.readthedocs.io/en/latest/OpenMM-interface.html#small-molecule-parameterizer .

If the molecule or residue with missing parameters is intended to always be in the QM-region then you don't need any bonded parameters (bond, angle, dihedrals) for that (since they would not be used anyway within a QM/MM scheme). You should simply define a basic nonbonded FF for that molecule, can be done using CHARMM files, Amber-files or OpenMM XML-files:
See discussion in these tutorials for how to do this with the OpenMM XML format:
https://ash.readthedocs.io/en/latest/Metalloprotein-I.html
https://ash.readthedocs.io/en/latest/Metalloprotein-II.html


> Lastly, I would really appreciate advice on how to proceed with a slightly convoluted missing atom case. I have an MD equilibrated system in AMBER that I would like to bring to ASH for QM/MM. The system was built with AMBER because MCPB.py protocol is needed for the metal binding active site. The system has a gas substrate that is missing, so I would have to add it before running ASH. Unfortunately, there are no straightforward protocols to directly add this substrate in AMBER tleap. I'd tried to port AMBER to CHARMM files to test ASH's built-in add atom functions, but there are no good ways to build all files needed for input.

Yeah, that is a bit tricky. The issue is that you have to modify the Amber PRMTOP file which is fairly complicated. For a very simple addition (adding 1 atom or something) it can be done manually if you understand where in the file you have to modify. For adding a whole new residue with bonded and nonbonded parameters you need some tool that is capable of intelligently modifying the PRMTOP file (or create a new one)
I would assume that there must be some option in Ambertools to do such a modification. However, I am not that familiar with the Amber way of doing things. Figuring this out using Ambertools or other Amber programs is probably your best shot.

Another option is to convert the Amber setup (PRMTOP) from Amber-format to OpenMM XML-format. I would avoid converting to CHARMM files like you tried.
Then you can just read in the converted system into ASH OpenMMTheory, using the converted XML-file and for the additional residue you can create a new XML-file (or add to the converted one). ASH can not do this but the ParmEd program is capable of this: https://parmed.github.io/ParmEd/html/index.html

Final option is to start almost from scratch but use the coordinate file (ideally PDB) from the Amber setup and have OpenMM find the FF files from the built-in Amber and then make the necessary additions.
Basically try to define an OpenMMTheory object within ASH like this:

`omm = OpenMMTheory(xmlfiles=["amber14-all.xml", "newfile.xml"], pdbfile="system.pdb")`

where system.pdb contains the coordinates for the whole system. amber14-all.xml is an Amber14 forcefield file (older versions are also there) that is built into OpenMM (it will find it automatically and hopefully should recognize most things in the PDB-file) and newfile.xml is a file that should contain special residue definitions and parameters. You will get an error complaining about missing residues and parameters initially, but in principle you can modify newfile.xml until things work out. Not easy though.

**Comment by Lisa-Phan** [comment]
Thank you Ragnar again for the response, 

I'd attempted to transfer the system to an OpenMM compatible XML with ParmEd. 
```
import parmed as pmd
from parmed.openmm.parameters import OpenMMParameterSet

INPCRD = 'amber_protein.inpcrd'
PRMTOP = 'amber_protein.prmtop'

#create a structure instance
amber = pmd.load_file(PRMTOP, INPCRD)

#write an OpenMM-compatible XML file
paramset = OpenMMParameterSet.from_structure(amber)
paramset.write('protein.xml')
```
This step proceeded without error message. When I load the file with ash

```
from ash import *

#equilibrated structure pdb
solvated_pdb = 'equilibrated_protein.pdb'

#amber forcefield xml, string that openmm can refer to internally?
amber_xml = 'amber14-all.xml'

#xml generated by parmed reading amber
openmm_protein_xml = 'protein.xml'

omm = OpenMMTheory(xmlfiles = [amber_xml, openmm_protein_xml], pdbfile = solvated_pdb)

```

There are some error messages on missing residue template. It seems like OpenMM doesn't like the non-standard amino acids residues that binds metal. I'd took a quick glance at the OpenMM github and it seems like there might be ways to add this, although not straightforward. 

Just as a clarification, a complete setup for ASH OpenMMTheory would have the PDB already containing protein + substrate, and information defined in xmlfiles would only be force and bond related constraints?

Best regards,
Lisa Phan

**Comment by RagnarB83** [comment]
Yes, the PDB-file should contain the whole system. The residue names and atom names (of the special things like substrate, metal etc.) should match the information in the XML-file. The special XML-file (i.e. your additions) should only contain FF parameters. For metal clusters and things that will either be frozen or described by QM, it is enough to only have LJ parameters and charges (i.e. nonbonded parameters) defined.

OK, good that Parmed succeeded. Presumably Parmed converted everything (i.e. the file contains FF definitions for all the amino acids etc.) so maybe you don't need to specify the amber14-all.xml at all. I would check that the file it created seems to contain everything.

But in case you do need to to specify the general Amber FF, maybe Amber14 is not appropriate for your Amber setup. Then other valid Amber FF versions have the corresponding XML-files: amber96.xml, amber10.xml, amber03.xml, amber96.xml, amber99sb.xml. These are files btw that exist within the OpenMM installation of your conda environment, OpenMM finds them automatically.

So if you are getting errors about non-standard amino acids you can do:
- You can start by deleting those residues from the PDB-file just to check if everything else works, apart from those residues. If it's just a few residues that are giving you problems then that can probably be dealt with, but if you are getting errors for almost everything then something is wrong and maybe the current approach is unlikely to succeed.
- Amino acids bound to metals are indeed likely to be an issue. It depends on how nonstandard those residues are. Deprotonated cysteines bound to Fe e.g. should work fine. On reason why things may not be working is if the PDB-file contains a CONECT line (end of PDB-file) that indicates that the residue-atom is bound to the metal ion while there is no such bond defined within the forcefield (bonds between atoms of different residues are usually not allowed). If this is the case then the problem can often be avoided by deleting the CONECT line. There will be no bond between those atoms at the forcefield level anymore but this would be fine if the metal ion and the residue is going to be in the in QM-region anyway.
- If it's a really unusual nonstandard amino acid then you could in principle define it manually within the XML-file. But presumably Parmed should already have done this ?

Hope this helps. Metalloprotein MM setup are tricky.
I can take a look also if you send me the files.

**Comment by Lisa-Phan** [comment]
Hi Ragnar, 

Thank you so much for the detailed comment. I'd probe around with the system a bit to see if any simple modification can get the workflow going. Unfortunately, none of what I'd tried is working. The error is likely not a CONECT issue, since none of the PDB used have the file. 

1. Regarding residue deletion, I am not sure yet how to remove a residue from PDB without causing the system to break. Do you have a recommendation on what to try?

2.  I think that a few error messages regarding missing template is related to amino acids around the metal site. I had thought it was a naming issue initially, but it seems like things are not so simple... 

3. Asides from the naming (GU2, HE2, etc) for the primary coordination amino acids, there are really no non-standard residues in the system. 

I'd put the things attempted in Zenodo, under this link. Please let me know if anything doesn't make sense. 
https://zenodo.org/records/13862695?token=eyJhbGciOiJIUzUxMiJ9.eyJpZCI6Ijc3ZWY2OWQxLTg4MjYtNDhhMi04NjQ2LTQ3OGIwMjVkZmNmYiIsImRhdGEiOnt9LCJyYW5kb20iOiI1YWMwMjA5ZTA2YTUxZTdiOGZkZmEwNGY4NDAzYjM5YSJ9.KnWu2YtI0Q6Q11mCX_UL2T9YHTBrN5GQAG5CxxFfMFbQJtVqWE8ha3eJef-lmQM5o-EV-FY56ev2Gz1I7mawkw

Thank you, 
Lisa Phan

**Comment by Lisa-Phan** [comment]
Hi Ragnar, 

This is wonderful! The files and syntax hints are much appreciated, and everything is running ok now. I know that the thread really deviated from the original question, so thank you for being patience with me. 

Best regards, 
Lisa Phan

**Comment by RagnarB83** [comment]
You're welcome, hope everything else works out. The BS part using ORCA should be a straightforward next step.

**Comment by RagnarB83** [comment]
You're welcome, hope everything else works out. The BS part using ORCA should be a straightforward next step.
